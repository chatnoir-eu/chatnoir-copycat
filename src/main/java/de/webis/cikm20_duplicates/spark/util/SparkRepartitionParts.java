package de.webis.cikm20_duplicates.spark.util;

import java.util.LinkedList;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

public class SparkRepartitionParts {
	
	public static void main(String[] args) {
		//String inputDir = "cikm2020/canonical-link-graph/cc-2017-04-canonical-urls";
//		String inputDir = "cikm2020/canonical-link-graph/cc-2017-04-sample-0.1";
		String inputDir = "cikm2020/canonical-link-graph/cc-2015-11-sample-0.1";
		
		try (JavaSparkContext context = context()) {
			for(String suffix: createPartSuffixes()) {
				JavaRDD<String> input = context.textFile(inputDir +"/part*" + suffix);
				input.repartition(5).saveAsTextFile(inputDir + "-delete-me-tmp-"+ suffix);
			}
			
			JavaRDD<String> completeInput = context.textFile(inputDir + "-delete-me-tmp-*/part*");
			completeInput.repartition(10000).saveAsTextFile(inputDir + "-new");
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/repartition-data");
	
		return new JavaSparkContext(conf);
	}

	public static List<String> createPartSuffixes() {
		List<String> ret = new LinkedList<>();
		
		for(int i=0; i < 10; i++) {
			for(int j=0; j<10; j++) {
				ret.add(i + "" + j);
			}
		}
		
		return ret;
	}
}
