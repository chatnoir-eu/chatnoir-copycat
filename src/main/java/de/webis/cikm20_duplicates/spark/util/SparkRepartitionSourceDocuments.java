package de.webis.cikm20_duplicates.spark.util;

import org.apache.hadoop.io.compress.BZip2Codec;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;

public class SparkRepartitionSourceDocuments {
	
	public static void main(String[] args) {
		try (JavaSparkContext context = context()) {
			String corpus = "cc-2015-11";
			JavaRDD<String> input = context.textFile("cikm2020/document-fingerprints-final/" + corpus +"*-jsonl");
			
			input.repartition(10000)
				.saveAsTextFile("cikm2020/document-fingerprints-final/" + corpus +"-jsonl.bzip2", BZip2Codec.class);
		}
	}
	
	private static JavaSparkContext context() {
		SparkConf conf = new SparkConf(true);
		conf.setAppName("cikm2020/repartition-data");

		return new JavaSparkContext(conf);
	}
}
