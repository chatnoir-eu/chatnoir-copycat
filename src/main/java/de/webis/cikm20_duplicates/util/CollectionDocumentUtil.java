package de.webis.cikm20_duplicates.util;

import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.time.temporal.ChronoUnit;

import org.apache.commons.io.IOUtils;

import de.webis.WebisUUID;
import de.webis.cikm20_duplicates.spark.SparkEnrichRelevanceTransferPairs;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.index.transform.JsoupStringTransform;
import lombok.SneakyThrows;
import net.jodah.failsafe.Failsafe;
import net.jodah.failsafe.RetryPolicy;

public class CollectionDocumentUtil {

	@SneakyThrows
	public static void main(String[] args) {
//		// S3: 1.0 (Should be valid)
//		String firstId = "clueweb12-0001wb-23-26353", secondId = "clueweb12-0005wb-24-06245";
		
//		// S3: 0.0 (Should be invalid)
//		String firstId = "clueweb12-0912wb-53-11472", secondId = "clueweb12-1000tw-11-07961";
		
		// S3: 0.0 (Should be invalid)
//		String firstId = "clueweb12-0303wb-10-02360", secondId = "clueweb12-0715wb-02-10722";

//		// S3: 0.0 (Should be invalid)
//		String firstId = "clueweb12-0303wb-35-15203", secondId = "clueweb12-0816wb-90-18920";
		
		// S3: 1.0 (Should be valid)
//		String firstId = "clueweb12-0406wb-47-22712", secondId = "clueweb12-1305wb-37-22239";
		
//		// S3: 1.0 (Should be valid)
//		String firstId = "clueweb12-0308wb-07-19989", secondId = "clueweb12-1505wb-58-23691";
		
//		// S3: 0.899 (Should be valid)
//		String firstId = "clueweb12-0810wb-02-19724", secondId = "clueweb12-0811wb-20-10663";
		
		
//		// S3: 0.99 (Should be valid)
//		String firstId = "clueweb12-0807wb-51-02126", secondId = "clueweb12-0810wb-43-15854";
		
//		// S3: 0.79 (should be invalid)
//		String firstId = "clueweb12-1313wb-20-00368", secondId = "clueweb12-1313wb-24-32697";
		
//		// S3: 0.94 (should be valid)
//		String firstId = "clueweb12-0805wb-13-09655", secondId = "clueweb12-0809wb-63-03994";
		
//		// S3: 0.97 (should be valid)
//		String firstId = "clueweb12-0810wb-79-00932", secondId = "clueweb12-0811wb-54-18364";
		
//		// S3: 0.93 (should be valid)
//		String firstId = "clueweb12-1600tw-18-12106", secondId = "clueweb12-1900tw-30-12707";
		
//		// S3: 0.79 (should be invalid)
//		String firstId = "clueweb12-1018wb-50-14351", secondId = "clueweb12-1019wb-33-29258";
		
//		System.out.println("Retrieve " + firstId + " located at: " + chatNoirURL(firstId) +"&plain");
//		CollectionDocument a = loadCollectionDocument(firstId);

//		System.out.println("Retrieve " + secondId + " located at: " + chatNoirURL(secondId) +"&plain");
//		CollectionDocument b = loadCollectionDocument(secondId);
		
//		System.out.println("Calculate S3-Score between " + firstId + " and " + secondId);
//		System.out.println(SparkEnrichRelevanceTransferPairs.s3Score(a,b));
		
		// {"equivalentDocuments": ["<urn:uuid:59a4637d-976f-4a5c-9861-cd5f3fc6c737>","<urn:uuid:be66ee99-8ee6-46ce-9e96-de6067eff76a>","<urn:uuid:ecffb099-540a-4176-8a97-7152a0576383>"],"hash":[-244842496, 47039, 11927603, 14439680]}
		//String firstId = "<urn:uuid:59a4637d-976f-4a5c-9861-cd5f3fc6c737>";
		//String secondId = "<urn:uuid:be66ee99-8ee6-46ce-9e96-de6067eff76a>";
		
		//{"equivalentDocuments": ["<urn:uuid:84606826-df14-4f0d-94cc-7d828f880949>","<urn:uuid:a1a75ce9-255f-4acd-83f9-30afd4fa6631>"],"hash":[-889782272, 50871, 14287080, 7504640]}
		// String firstId = "<urn:uuid:84606826-df14-4f0d-94cc-7d828f880949>";
		// String secondId = "<urn:uuid:a1a75ce9-255f-4acd-83f9-30afd4fa6631>";
		
		//{"equivalentDocuments": ["<urn:uuid:aa833e3e-0033-4937-b499-14e26d972404>","<urn:uuid:fb1ec340-b003-4dad-aa02-9a3ebe34b75f>"],"hash":[1527775232, 344, 13172802, 9114880]}
		// String firstId = "<urn:uuid:aa833e3e-0033-4937-b499-14e26d972404>";
		// String secondId = "<urn:uuid:fb1ec340-b003-4dad-aa02-9a3ebe34b75f>";
		
		//{"equivalentDocuments": ["<urn:uuid:177caa19-fd12-48dd-b85f-aa9151e1048e>","<urn:uuid:c74bf9bb-c949-47a1-bd4c-967e528a3e0d>","<urn:uuid:fd3341db-5dbb-45fe-b5c3-7aa1f91527a6>"],"hash":[367919104, 13594, 1900705, 413184]}
		String firstId = "<urn:uuid:177caa19-fd12-48dd-b85f-aa9151e1048e>";
		String secondId = "<urn:uuid:fd3341db-5dbb-45fe-b5c3-7aa1f91527a6>";
		
		System.out.println(chatNoirURL("commoncrawl", firstId, "cc1704"));
		CollectionDocument a = loadCollectionDocument(firstId, new URL(chatNoirURL("commoncrawl", firstId, "cc1704")));
		System.out.println(chatNoirURL("commoncrawl", secondId, "cc1704"));
		CollectionDocument b = loadCollectionDocument(secondId, new URL(chatNoirURL("commoncrawl", secondId, "cc1704")));
		
		System.out.println(SparkEnrichRelevanceTransferPairs.s3Score(a,b));
	}
	
	public static final RetryPolicy<String> RETRY_FINAL = new RetryPolicy<String>()
			.handle(Exception.class)
			.withBackoff(3, 60, ChronoUnit.SECONDS)
			.withMaxRetries(3);	
	
	@SneakyThrows
	public static CollectionDocument loadCollectionDocument(String documentId) {
		return loadCollectionDocument(new URL(chatNoirURL(documentId)), documentId, RETRY_FINAL);
	}
	

	public static CollectionDocument loadCollectionDocument(String documentId, URL chatNoirUrl) {
		return loadCollectionDocument(chatNoirUrl, documentId, RETRY_FINAL);
	}
	
	public static CollectionDocument loadCollectionDocument(URL url, String documentId, RetryPolicy<String> retryPolicy) {
		String documentContent = loadRawDocumentFromChatnoir(url, retryPolicy);
		
		JsoupStringTransform stringTransform = new JsoupStringTransform();
		return CollectionDocument.collectionDocument(stringTransform.apply(documentContent), documentId);
	}
	
	public static String loadRawDocumentFromChatnoir(URL url, RetryPolicy<String> retryPolicy) {
		return Failsafe.with(retryPolicy).get(() -> IOUtils.toString(url, StandardCharsets.UTF_8));
	}

	private static String webisUUID(String documentId) {
		return webisUUID(longChatNoirId(documentId), documentId);
	}
	
	private static String webisUUID(String prefix, String documentId) {
		return new WebisUUID(prefix)
				.generateUUID(documentId).toString();
	}
	
	private static String longChatNoirId(String documentId) {
		if(documentId.startsWith("clueweb09")) {
			return "clueweb09";
		} else if (documentId.startsWith("clueweb12")) {
			return "clueweb12";
		}
		
		throw new RuntimeException("ID '" + documentId + "' is not supported.");
	}
	
	private static String shortChatNoirId(String documentId) {
		if (documentId.startsWith("clueweb09")) {
			return "cw09";
		} else if (documentId.startsWith("clueweb12")) {
			return "cw12";
		}
		
		throw new RuntimeException("ID '" + documentId + "' is not supported.");
	}
	
	public static String chatNoirURL(String documentId) {
		return "https://chatnoir.eu/cache?uuid=" + webisUUID(documentId) + "&index="+ shortChatNoirId(documentId) +"&raw";
	}
	
	public static String chatNoirURL(String prefix, String documentId, String index) {
		return "https://chatnoir.eu/cache?uuid=" + webisUUID(prefix, documentId) + "&index="+ index +"&raw";
	}
}
