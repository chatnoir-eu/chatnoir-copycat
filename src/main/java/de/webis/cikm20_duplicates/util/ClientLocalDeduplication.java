package de.webis.cikm20_duplicates.util;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.curator.shaded.com.google.common.collect.Iterators;
import org.apache.htrace.shaded.fasterxml.jackson.databind.ObjectMapper;

import com.google.common.collect.ImmutableList;

import de.aitools.ir.fingerprinting.representer.Hash;
import de.webis.cikm20_duplicates.spark.SparkCreateDeduplicationCandidates.DeduplicationUnit;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import scala.Tuple2;

public class ClientLocalDeduplication {

	public static List<String> dedup(Iterable<Tuple2<Integer, DeduplicationUnit>> group) {
		List<String> ret = new ArrayList<>();
		Map<String, List<Tuple2<Integer, DeduplicationUnit>>> equals = sortedList(group);

		for(List<Tuple2<Integer, DeduplicationUnit>> equal: equals.values()) {
			ret.addAll(equalPairs(equal));
		}
		
		for(List<DeduplicationUnit> toDedup: deduplicationPairs(equals)) {
			for(String bla : fullDeduplication(toDedup)) {
				if(!ret.contains(bla)) {
					ret.add(bla);
				}
			}
		}
		
		return ret;
	}
	
	public static Collection<List<DeduplicationUnit>> deduplicationPairs(Map<String, List<Tuple2<Integer, DeduplicationUnit>>> equals) {
		Map<Integer, List<DeduplicationUnit>> ret = new LinkedHashMap<>();
		
		if(equals.size() < 100000) {
			List<DeduplicationUnit> tmpRet = new ArrayList<>();
			for(List<Tuple2<Integer, DeduplicationUnit>> equal: equals.values()) {
				if(equal.size() < 1) {
					continue;
				}
				
				tmpRet.add(equal.get(0)._2);
			}
			
			return Arrays.asList(tmpRet);
		}
		
		for(List<Tuple2<Integer, DeduplicationUnit>> equal: equals.values()) {
			if(equal.size() < 1) {
				continue;
			}
			
			Tuple2<Integer, DeduplicationUnit> i = equal.get(0);
			List<Integer> rekursiveInts = HashTransformationUtil.removeIntFromUnderlyingBitArray(i._2.getHashParts(), i._1());
			for(int recursivePart: rekursiveInts) {
				if(!ret.containsKey(recursivePart)) {
					ret.put(recursivePart, new ArrayList<>());
				}
				
				ret.get(recursivePart).add(i._2);
			}
		}
		
		return ret.values();
	}
	
	public static List<String> fullDeduplication(List<DeduplicationUnit> orig) {
		List<String> ret = new LinkedList<>();
		List<Tuple2<String, byte[]>> idToHash = new ArrayList<>(orig.size());
		
		for(DeduplicationUnit d: orig) {
			idToHash.add(new Tuple2<>(d.getId(), HashTransformationUtil.integersToHash(d.getHashParts())));
		}
		
		for(int i=0; i<idToHash.size(); i++) {
			Tuple2<String, byte[]> left = idToHash.get(i);

			for(int j=i+1; j< idToHash.size(); j++) {
				Tuple2<String, byte[]> right = idToHash.get(j);
				
				int hemming = Hash.getHammingDistance(left._2(), right._2());
				if(hemming <= 3) {
					ret.add(nearDuplicate(left._1(), right._1(), hemming));
				}
			}
		}
		
		return ret;
	}

	private static Collection<String> equalPairs(List<Tuple2<Integer, DeduplicationUnit>> equal) {
		List<String> ret = new ArrayList<>();
		
		for(int i= 1; i< equal.size(); i++) {
			ret.add(nearDuplicate(equal.get(0)._2.getId(), equal.get(i)._2.getId(), 0));
		}
		
		return ret;
	}
	
	private static String nearDuplicate(String id1, String id2, int hemming) {
		return "{\"firstId\":\"" + id1 + "\",\"secondId\":\"" + id2 + "\",\"hemmingDistance\":"+ hemming +"}";
	}


	private static void failIfElementsAreNotAllFromTheSameGroup(Iterable<Tuple2<Integer, DeduplicationUnit>> group) {
		Set<Integer> ret = new HashSet<>();
		Iterators.transform(group.iterator(), i -> i._1()).forEachRemaining(i -> ret.add(i));
		
		if(ret.size() > 1) {
			throw new IllegalArgumentException("");
		}
	}
	
	public static Map<String, List<Tuple2<Integer, DeduplicationUnit>>> sortedList(Iterable<Tuple2<Integer, DeduplicationUnit>> group) {
		List<Tuple2<Integer, DeduplicationUnit>> tmp = ImmutableList.copyOf(group.iterator());
		tmp = new ArrayList<>(tmp);
		Collections.sort(tmp, (a,b) -> a._2.getId().compareTo(b._2.getId()));
		failIfElementsAreNotAllFromTheSameGroup(tmp);
		
		Map<String, List<Tuple2<Integer, DeduplicationUnit>>> ret = new LinkedHashMap<>();
		for(Tuple2<Integer, DeduplicationUnit> t : tmp) {
			String key = t._2().getHashParts().toString();
			
			if(!ret.containsKey(key)) {
				ret.put(key, new LinkedList<>());
			}
			
			ret.get(key).add(t);
		}
		
		
		return ret;
	}
	
	@Data
	@NoArgsConstructor
	@AllArgsConstructor
	public static class DeduplicationTask {
		private List<DeduplicationUnit> entries;
		
		@Override
		@SneakyThrows
		public String toString() {
			return new ObjectMapper().writeValueAsString(this);
		}
		
		@SneakyThrows
		public static DeduplicationTask fromString(String value) {
			return new ObjectMapper().readValue(value, DeduplicationTask.class);
		}
	}

	public static java.util.Iterator<String> workingPackages(Iterable<Tuple2<Integer, DeduplicationUnit>> group) {
		Map<String, List<Tuple2<Integer, DeduplicationUnit>>> equals = sortedList(group);
		Collection<List<DeduplicationUnit>> ret = deduplicationPairs(equals);
		
		return com.google.common.collect.Iterators.transform(
			ret.iterator(), 
			i -> new DeduplicationTask(i).toString()
		);
	}
}
