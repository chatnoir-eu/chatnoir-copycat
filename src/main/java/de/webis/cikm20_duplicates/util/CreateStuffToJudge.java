package de.webis.cikm20_duplicates.util;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.databind.ObjectMapper;

import de.webis.cikm20_duplicates.spark.SparkCalculateCanonicalLinkGraphEdgeLabels.CanonicalLinkGraphEdge2;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import lombok.SneakyThrows;

public class CreateStuffToJudge {
	private static final String
		DIR = "src/main/resources/canonical-link-edges-human-judgments/",
		
		//CREATED BY: de.webis.cikm20_duplicates.spark.eval.SparkSampleS3EdgesPerBin
		INPUT = DIR + "cw12-sampled-edges.jsonl";
	
	@SneakyThrows
	public static void main(String[] args) {
		for(String line: Files.readAllLines(Paths.get(INPUT))) {
			persistExamples(line);
		}
	}
	
	@SneakyThrows
	@SuppressWarnings({ "unchecked", "rawtypes" })
	private static void persistExamples(String json) {
		Map<String, Object> m = new ObjectMapper().readValue(json, Map.class); 
		
		if(!m.containsKey("s3Bucket") || m.get("s3Bucket") == null) {
			return;
		} else {
			System.out.println("Bucket: " + m.get("s3Bucket"));
		}
		
		List edges = (List) m.get("sample-edges");
		edges = edges.subList(0, 10);
		
		for(Object e: edges) {
			String eStr = new ObjectMapper().writeValueAsString(e);
			CanonicalLinkGraphEdge2 eVal = CanonicalLinkGraphEdge2.fromString(eStr);
			
			persistExample(eVal);
		}
	}
	
	@SneakyThrows
	private static void persistExample(CanonicalLinkGraphEdge2 bla) {
		CollectionDocument src = bla.getFirstDoc().getDoc();
		CollectionDocument target = bla.getSecondDoc().getDoc();
		
		Path dir = Paths.get(DIR + "judgments-by-daniel/" + src.getId() + "__to__" + target.getId());
		dir.toFile().mkdirs();
		
		Files.write(dir.resolve("metadata"), metadata(src, target, bla.getCanonicalLink(), bla.getS3score()).getBytes(StandardCharsets.UTF_8));
		Files.write(dir.resolve("src-doc"), src.getFullyCanonicalizedContent().getBytes(StandardCharsets.UTF_8));
		Files.write(dir.resolve("target-doc"), target.getFullyCanonicalizedContent().getBytes(StandardCharsets.UTF_8));
	}
	
	private static String metadata(CollectionDocument a, CollectionDocument b, String canonicalUrl, double s3) {
		return "Canonical-Link: " + canonicalUrl +"\n" +
				"First: "+ CollectionDocumentUtil.chatNoirURL(a.getId()) +"&plain\n" + 
				"Second: "+ CollectionDocumentUtil.chatNoirURL(b.getId()) +"&plain\n" +
				"S3: " + s3 +"\n";
	}
}
