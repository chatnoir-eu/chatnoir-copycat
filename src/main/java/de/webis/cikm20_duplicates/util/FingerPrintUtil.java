package de.webis.cikm20_duplicates.util;

import java.io.Serializable;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import de.webis.trec_ndd.trec_collections.CollectionDocument;
import de.webis.trec_ndd.util.NGramms;
import info.debatty.java.lsh.MinHash;
import lombok.experimental.UtilityClass;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
@UtilityClass
public class FingerPrintUtil {
	public static interface Fingerprinter<T extends Comparable<T>> extends Serializable {
		public List<T> fingerprint(CollectionDocument doc);
		
		public double similarity(List<T> a, List<T> b);
	}

	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> minHashFingerPrinting(long seed) {
		int dict_size = 1;
		int size = 12;
		MinHash minHash = new MinHash(size, dict_size, seed);
		
		return new Fingerprinter<Integer>() {
			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				return IntStream.of(minHash.signature(docToElementSet(doc))).boxed().collect(Collectors.toList());
			}

			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				int[] aArray = a.stream().mapToInt(i -> i).toArray();
				int[] bArray = b.stream().mapToInt(i -> i).toArray();
				
				return minHash.similarity(aArray, bArray);
			}
		};
	}
	
	@SuppressWarnings("serial")
	public static Fingerprinter<Integer> simHashFingerPrinting(int bitsInSimHash, int k) {
		return new Fingerprinter<Integer>() {

			@Override
			public List<Integer> fingerprint(CollectionDocument doc) {
				return Arrays.asList(1,1,1,1);
			}

			@Override
			public double similarity(List<Integer> a, List<Integer> b) {
				return 1.0;
			}
		};
	}
	
	private static Set<Integer> docToElementSet(CollectionDocument doc) {
		return NGramms.build8Gramms(doc.getFullyCanonicalizedContent()).stream()
				.map(i -> i.getMd5Hash().hashCode())
				.collect(Collectors.toSet());
	}
}
