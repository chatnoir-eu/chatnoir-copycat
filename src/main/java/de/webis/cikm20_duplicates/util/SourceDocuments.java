package de.webis.cikm20_duplicates.util;

import java.io.Serializable;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Set;

import org.apache.htrace.shaded.fasterxml.jackson.databind.ObjectMapper;

import de.webis.trec_ndd.trec_collections.SharedTask;
import de.webis.trec_ndd.trec_collections.CollectionConfiguration.TrecCollections;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import de.webis.trec_ndd.trec_collections.SharedTask.TrecSharedTask;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.SneakyThrows;
import lombok.experimental.UtilityClass;
import lombok.experimental.Wither;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
@UtilityClass
public final class SourceDocuments {
	private static final List<TrecSharedTask> TRACKS = Arrays.asList(
			TrecSharedTask.WEB_2009, TrecSharedTask.WEB_2010,
			TrecSharedTask.WEB_2011, TrecSharedTask.WEB_2012,

			TrecSharedTask.WEB_2013, TrecSharedTask.WEB_2014,

			TrecSharedTask.SESSION_2010, TrecSharedTask.SESSION_2011,
			TrecSharedTask.SESSION_2012,

			TrecSharedTask.SESSION_2013, TrecSharedTask.SESSION_2014,
			
			TrecSharedTask.ENTITY_2009, TrecSharedTask.ENTITY_2010,
			TrecSharedTask.ENTITY_2011
	);
	
	public static final List<SourceDocument> ALL_DOCS_FOR_WHICH_DUPLICATES_SHOULD_BE_SEARCHED = getAllDocumentsForWhichDuplicatesShouldBeSearched();
	
	@Data
	@Wither
	@NoArgsConstructor
	@AllArgsConstructor
	@SuppressWarnings("serial")
	public static class CollectionDocumentWithTopics implements Serializable {
		private CollectionDocument doc;
		private List<String> topics;
		
		@Override
		@SneakyThrows
		public String toString() {
			return new ObjectMapper().writeValueAsString(this);
		}
		
		@SneakyThrows
		public static CollectionDocumentWithTopics fromString(String source) {
			return new ObjectMapper().readValue(source, CollectionDocumentWithTopics.class);
		}
	}
	
	@Data
	@Wither
	@AllArgsConstructor
	public static class SourceDocument {
		private final TrecSharedTask task;
		private final TrecCollections collection;
		private final String documentId;
		private final String topic;
		
		public String toString() {
			return collection.name() + "::" + task.name() + "::" + topic + "::" + documentId;
		}
	}

	private static List<SourceDocument> getAllDocumentsForWhichDuplicatesShouldBeSearched() {
		List<SourceDocument> ret = new LinkedList<>();
		for (SharedTask task : TRACKS) {
			Set<String> topics = task.documentJudgments().getData().keySet();

			for (String topic : topics) {
				Set<String> judgedDocs = task.documentJudgments().getData().get(topic).keySet();
				for (String doc : judgedDocs) {
					ret.add(new SourceDocument(
							(TrecSharedTask) task, collectionForTask(task),
							doc, topic
					));
				}
			}
		}

		Collections.sort(ret, (a,b) -> a.toString().compareTo(b.toString()));
		
		return Collections.unmodifiableList(ret);
	}
	
	public static Map<TrecSharedTask, List<SourceDocument>> taskToDocuments() {
		Map<TrecSharedTask, List<SourceDocument>> ret = new LinkedHashMap<>();
		
		for(SourceDocument doc : ALL_DOCS_FOR_WHICH_DUPLICATES_SHOULD_BE_SEARCHED) {
			if(!ret.containsKey(doc.getTask())) {
				ret.put(doc.getTask(), new LinkedList<>());
			}
			
			ret.get(doc.getTask()).add(doc);
		}
		
		return ret;
	}
	
	public static Map<String, Set<String>> topicsToDocumentIds() {
		Map<String, Set<String>> ret = new LinkedHashMap<>();
		
		for(SourceDocument doc : ALL_DOCS_FOR_WHICH_DUPLICATES_SHOULD_BE_SEARCHED) {
			String topic = doc.collection.name() + "::" + doc.task.name() + "::" + doc.topic;
			if(!ret.containsKey(topic)) {
				ret.put(topic, new HashSet<>());
			}
			
			ret.get(topic).add(doc.documentId);
		}
		
		return ret;
	}
	
	private static TrecCollections collectionForTask(SharedTask task) {
		if(Arrays.asList(TrecSharedTask.WEB_2009, TrecSharedTask.WEB_2010,
			TrecSharedTask.WEB_2011, TrecSharedTask.WEB_2012).contains(task)) {
			return TrecCollections.CLUEWEB09;
		} else if(Arrays.asList(TrecSharedTask.WEB_2013, TrecSharedTask.WEB_2014).contains(task)) {
			return TrecCollections.CLUEWEB12;
		} else if(Arrays.asList(TrecSharedTask.SESSION_2010, TrecSharedTask.SESSION_2011,
				TrecSharedTask.SESSION_2012).contains(task)) {
			return TrecCollections.CLUEWEB09;
		} else if(Arrays.asList(TrecSharedTask.SESSION_2013, TrecSharedTask.SESSION_2014).contains(task)) {
			return TrecCollections.CLUEWEB12;
		} else if (Arrays.asList(TrecSharedTask.ENTITY_2009, TrecSharedTask.ENTITY_2010,
			TrecSharedTask.ENTITY_2011).contains(task)) {
			return TrecCollections.CLUEWEB09;
		}
		
		throw new RuntimeException("fixme");
	}
}
