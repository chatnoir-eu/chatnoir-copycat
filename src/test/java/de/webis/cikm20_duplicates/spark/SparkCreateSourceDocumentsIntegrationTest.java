package de.webis.cikm20_duplicates.spark;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Iterator;
import java.util.List;

import org.apache.spark.api.java.JavaRDD;
import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Test;

import com.holdenkarau.spark.testing.SharedJavaSparkContext;

import de.webis.cikm20_duplicates.util.SourceDocuments.CollectionDocumentWithTopics;
import de.webis.trec_ndd.trec_collections.AnseriniCollectionReader;
import de.webis.trec_ndd.trec_collections.CollectionDocument;
import io.anserini.collection.ClueWeb09Collection.Document;

/**
 * 
 * @author Maik Fr√∂be
 *
 */
public class SparkCreateSourceDocumentsIntegrationTest extends SharedJavaSparkContext {
	@Test
	public void testWithEmptyDocuments() {
		AnseriniCollectionReader<Document> acr = new DummyAnseriniCollectionReader();
		long expected = 0;
		
		JavaRDD<CollectionDocumentWithTopics> rdd = SparkCreateSourceDocuments.transformAllImportantDocuments(jsc(), acr);
		
		Assert.assertEquals(expected, rdd.count());
	}
	
	@Test
	public void testWithOnlyUnimportantDocuments() {
		AnseriniCollectionReader<Document> acr = new DummyAnseriniCollectionReader(
			doc("a"), doc("b"), doc("c")
		);
		long expected = 0;
		
		JavaRDD<CollectionDocumentWithTopics> rdd = SparkCreateSourceDocuments.transformAllImportantDocuments(jsc(), acr);
		
		Assert.assertEquals(expected, rdd.count());
	}
	
	@Test
	public void testWithSomeImportantAndSomeUnimportantDocuments() {
		AnseriniCollectionReader<Document> acr = new DummyAnseriniCollectionReader(
				doc("a"), doc("b"), 
				doc("clueweb09-en0008-02-29970"), // judged for topic 50 of the web track
				doc("c")
		);
		JavaRDD<CollectionDocumentWithTopics> rdd = SparkCreateSourceDocuments.transformAllImportantDocuments(jsc(), acr);
		List<String> actual = sorted(rdd);

		Approvals.verifyAsJson(actual);
	}
	
	@Test
	public void testMultipleAcrsWithSomeImportantAndSomeUnimportantDocuments() {
		AnseriniCollectionReader<Document> acr1 = new DummyAnseriniCollectionReader(
				doc("a"), doc("b"), 
				doc("clueweb09-en0008-02-29970"), // judged for topic 50 of the web track
				doc("c")
		);
		AnseriniCollectionReader<Document> acr2 = new DummyAnseriniCollectionReader(
				doc("a"), doc("b"), doc("c")
		);
		AnseriniCollectionReader<Document> acr3 = new DummyAnseriniCollectionReader(
				doc("d"), doc("b"), 
				doc("clueweb12-1800tw-04-16339") // judged for topic 60 of the session track 
		);
		
		JavaRDD<CollectionDocumentWithTopics> rdd = SparkCreateSourceDocuments.transformAllImportantDocuments(jsc(), acr1, acr2, acr3);
		List<String> actual = sorted(rdd);
		
		Approvals.verifyAsJson(actual);
	}
	
	@SuppressWarnings("serial")
	public static class DummyAnseriniCollectionReader extends AnseriniCollectionReader<Document> {

		private final ArrayList<CollectionDocument> docs;
		
		public DummyAnseriniCollectionReader(CollectionDocument...docs) {
			super(null);
			this.docs = new ArrayList<>(Arrays.asList(docs));
		}
		
		@Override
		public List<String> segmentPaths() {
			return Arrays.asList("a");
		}
		
		@Override
		public Iterator<CollectionDocument> collectionDocumentsInPath(String segmentPath) {
			return docs.iterator();
		}
	}
	
	private static CollectionDocument doc(String id) {
		return new CollectionDocument(id, "content of " + id, "fullyCanonicalizedContent of " + id);
	}
	
	private static List<String> sorted(JavaRDD<CollectionDocumentWithTopics> rdd) {
		List<String> ret = rdd.map(i -> i.toString()).collect();
		ret = new ArrayList<>(ret);
		Collections.sort(ret);
		
		return ret;
	}
}
