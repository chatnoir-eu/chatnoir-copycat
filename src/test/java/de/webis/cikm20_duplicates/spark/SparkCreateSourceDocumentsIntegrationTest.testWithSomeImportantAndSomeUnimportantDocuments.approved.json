[
  "({\"id\":\"clueweb09-en0008-02-29970\",\"content\":\"content of clueweb09-en0008-02-29970\",\"fullyCanonicalizedContent\":\"fullyCanonicalizedContent of clueweb09-en0008-02-29970\"},[CLUEWEB09::WEB_2009::50])"
]